translator/trace/zipkinv1_to_protospan.go

```diff
@@ -88,11 +90,24 @@ func ZipkinV1JSONBatchToOCProto(blob []byte) ([]*agenttracepb.ExportTraceService
 		return nil, errors.WithMessage(err, msgZipkinV1JSONUnmarshalError)
 	}
 
+	return zipkinToOCProto(func(i int) (*tracepb.Span, *annotationParseResult, error) {
+		if i >= len(zSpans) {
+			return nil, nil, errNoMoreSpans
+		}
+
+		return zipkinV1ToOCSpan(zSpans[i])
+	})
+}
+
+func zipkinToOCProto(toOCSpan func(i int) (*tracepb.Span, *annotationParseResult, error)) ([]*agenttracepb.ExportTraceServiceRequest, error) {
```

Comment:
Why is this function's idiom an iterator on an index instead of just iterating per span?
This new change makes for a non-pure-like function. I understand the need for specialized
processing with the toOCSpan but how about
```go
func zipkinToOCProto(zSpans []*zipkin.Span, toOCSpan func(zs *zipkincore.Span) (*tracepb.Span, *annotationParseResult, error)) ([]*agenttracepb.ExportTraceServiceRequest, error) {
       for _, zSpan := range zSpans {
             ocSpan, parsedAnnotations, err := zipkinV1ToOCSpan(zSpan)
             ...
      }
}
```

and that when used will just be simple matter of
```go
return zipkinToOCProto(zSpans, zipkinV1ToOCSpan)
```
instead of the repeated `if i >= len(zSpans)`

Reply:
In order to do that there is a conversion to zipkincore.Span that the JSON path doesn't need (per related comment above). The alternative seems to eager generate the slice with oc span and respective parsed annotation.

Reply:
or a function that instead takes an interface, and do the cast inside...

Reply:
in the end I wasn't satisfied with the two alternatives

Reply:
>In order to do that there is a conversion to zipkincore.Span that the JSON path doesn't need (per related comment above). The alternative seems to eager generate the slice with oc span and respective parsed annotation.

if you have `toOCSpan` as specialized function that just takes in a `zipkincore.Span` that's the equivalent of taking an index and then looking it up and it even saves us from having to index on our own and then compare against the length of the slice then return `errNoMoreSpans`.
I still don't get the rationale here for the `for i := 0; ; i++` and then a mapping function that takes in an index yet we could simply iterate over each span.

Reply:
`for i := 0; ; i++` and the `errNoMoreSpans` will be removed (something that ended in between two ideas). 

> if you have toOCSpan as specialized function that just takes in a zipkincore.Span 

The usage of `toOCSpan` with an index is to avoid converting from `zipkinV1Span` to `zipkincore.Span` (or vice-versa). The simplest alternative to get rid of it is to do the all transformations (to the triples) before going to the "re-used" loop, i.e.: `zipkinToOCProto` it will allocate an extra array per call and will have very similar loops for v1 json and thrift functions but likely very minimal impact on performance overall (the function itself should be much more expensive than the extra array).

Reply:
confirmed: 1 extra allocation and ~1% longer per benchmark, but the overall time measured is around 10ms. It is acceptable.

Reply:
>The usage of toOCSpan with an index is to avoid converting from zipkinV1Span to zipkincore.Span (or vice-versa). The simplest alternative to get rid of it is to do the all transformations (to the triplets) before going to the "re-used" loop, i.e.: zipkinToOCProto. It will allocate an extra array per call and will have very similar loops for v1 json and thrift functions but likely very minimal impact on performance overall (the function itself should be much more expensive than the extra array).

Currently I think let's go for clarity before optimization and then can optimize as needed. Thank you for making the changes for clearer code :)